Improve typeof operand highlighting (WIP)

dkearns opened the pull request:
Currently the `typeof` operand and parens are all matched as `csTypeofStatement`.

This looks very odd to my eye.  I'd expect the parens to be in the default foreground colour and wouldn't expect the operand to get special treatment either beyond the normal type highlighting.

Someone has obviously gone to some trouble to implement this behaviour though so I may be missing something obvious.  See: https://github.com/nickspoons/vim-cs/commit/5ffc8074b7fe062bea573cc648cde8d717997a38

The paren is also currently required to immediately follow the keyword which I assume is just a bug.

This commit allows the following paren with intervening whitespace (including empty lines which is probably overkill) and marks anything following it that's not an opening paren as an error.  It matches the parens as the default paren group (csParens) and highlights the operand specially as is currently done.

Personally, I'd probably just go with the following which highlights the operand type normally.

```vim
syn keyword csTypeOf        typeof nextgroup=csTypeOfOperand skipwhite
syn region  csTypeOfOperand matchgroup=csParens start="(" end=")" transparent contains=csType
```

dkearns committed:
Improve typeof operand highlighting

Don't highlight parens specially with the operand highlight group.

Allow whitespace between the keyword and opening paren.

nickspoons commented:
Yeah this is much better. Although why include `transparent`? The `csTypeOfOperand` is linked to `TypeDef` but that highlight isn't applied due to the `transparent`.

dkearns commented:
Sorry, `transparent` should have been removed from the compromise implementation I was suggesting.

Having dug a little deeper, I see that within the syntax file that this implementation originated they were also matching type names so it made a bit more sense to match the whole operand and then refine it with these contained items.

Unless you're keen on enforcing the presence of the `(` before `typeof` is highlighted or matching a following non-open-paren as an error.  I can't seen any reason not to just match `typeof` as something like `syn keyword csOperator typeof` and be done with it.   There's no real need to do anything special with the operand.

On a loosely related note I might look at adding attribute highlighting soon.

nickspoons commented:
This is good, thanks.

> I can't seen any reason not to just match typeof as something like syn keyword csOperator typeof and be done with it. There's no real need to do anything special with the operand.

While I see your point, this is a fairly rigid syntax element that is easy enough to capture so it can be highlighted as desired. Having it in groups as we have here makes it simple to modify highlights as needed, and also to prevent collisions with other syntax elements such as method calls. The fact that others have gone out of their way to capture it in the past indicates that this is something people want.

> On a loosely related note I might look at adding attribute highlighting soon.

Go nuts! Let me know when you've finished with the current "batch" of updates, and then I'll send the updated files to Bram.

nickspoons commented:
Hmm. I just added github actions to make the Vader tests clearer, and it immediately pointed out that this has caused an error.

I sort of wonder if it's a Vader bug though. When I check the highlight group of the character in question (the first `<` in [test/operators.vader](https://github.com/nickspoons/vim-cs/blob/master/test/operators.vader#L4), it appears to be `csOpSymbols` as expected. However the Vader function `echo vader#helper#syntax_at()` returns `csTypeOfOperand`.

Whether or not this is a vader issue, it is peculiar that this highlight group is applied here. Can we tighten it up so that this group only ever applies after `typeof`?

nickspoons commented:
What do you think of this?

```vim
syn keyword	csTypeOf	typeof contained
syn region	csTypeOfOperand	matchgroup=csParens start="(" end=")" contains=csType contained
syn match	csTypeOfStatement	"typeof\s*(\s*\w\+\s*)" display contains=csTypeOf,csTypeOfOperand
```

This way there is no highlighting of `typeof` until it is followed by a valid pattern, which is what you described earlier. Perhaps it's too much and we should just be highlighting the keyword, as you've described.

@idbrii you introduced the `csTypeOfStatement`, do you have strong feelings about this?

dkearns commented:
> Hmm. I just added github actions to make the Vader tests clearer, and it immediately pointed out that this has caused an error.

Sorry, my (WIP) heading tag usually means I haven't bothered testing it thoroughly yet because it's still being worked out.  Fixed in #26.

dkearns commented:
> What do you think of this?
> 
> ```viml
> syn keyword	csTypeOf	typeof contained
> syn region	csTypeOfOperand	matchgroup=csParens start="(" end=")" contains=csType contained
> syn match	csTypeOfStatement	"typeof\s*(\s*\w\+\s*)" display contains=csTypeOf,csTypeOfOperand
> ```
> 
> This way there is no highlighting of `typeof` until it is followed by a valid pattern, which is what you described earlier. Perhaps it's too much and we should just be highlighting the keyword, as you've described.

From a user perspective I find highlighting dependent on unnecessary look-ahead to be as annoying as bad error highlighting.

The error is actually the missing `(` not `typeof` being invalid in that context.  E.g., " error CS1003: Syntax error, '(' expected "

idbrii commented:
Here's some tricky sample code using typeof.
```cs
namespace Example
{
    [RequireComponent(typeof(Collider))]
    public class UsingAttributes : MonoBehaviour
    {
        [AOT.MonoPInvokeCallback(typeof(System.DEBUG_CALLBACK))]
        static Result Callback()
        {
            return Create(typeof(Result));
        }
    }
}
```

@nickspoons the change [you suggest here](https://github.com/nickspoons/vim-cs/pull/23#issuecomment-984101601) fails to highlight the typeof in the function's attribute.

It looks like [the latest version of syntax](https://github.com/nickspoons/vim-cs/blob/58c69195807ed24af914d9b049b5a87a4cced38a/syntax/cs.vim) fails to highlight either attribute correctly.

----

As for how I'd like it to work: Obviously `typeof` should show as Keyword. I think `typeof Collider` shouldn't highlight correctly because it's an error. Showing as an actual error highlight sounds like a good idea if it can never be wrong.

I like that `Collider` inside `typeof(Collider)` is a `Typedef` instead of a `Type`. Since c# doesn't have typedefs, I like giving the option of more more colour diversity. (I think most colorschemes make `Typedef` and `Type` look the same, but [mine](https://github.com/idbrii/vim-sandydune) does not.) I prefer that the contents of the typeof is coloured because it groups the statement as a whole in my eyes.

nickspoons commented:
> Sorry, my (WIP) heading tag usually means I haven't bothered testing it thoroughly yet 

Oh yeah I was a bit quick there, apoligies.

nickspoons commented:
Thanks for the feedback, @idbrii. I think the latest fix from @dkearns covers everything you've raised.
